modules/distributor/forwarder/otlpgrpc/forwarder.go

```diff
@@ -0,0 +1,174 @@
+package otlpgrpc
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/go-kit/log"
+	grpcmw "github.com/grpc-ecosystem/go-grpc-middleware"
+	"github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/weaveworks/common/middleware"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/collector/pdata/ptrace/ptraceotlp"
+	"go.uber.org/multierr"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+type Forwarder struct {
+	cfg         Config
+	logger      log.Logger
+	connections map[string]*grpc.ClientConn
+	clients     map[string]ptraceotlp.Client
+	initialized bool
+	mu          *sync.RWMutex
+}
+
+func NewForwarder(cfg Config, logger log.Logger) (*Forwarder, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate config: %w", err)
+	}
+
+	return &Forwarder{
+		cfg:         cfg,
+		logger:      logger,
+		connections: make(map[string]*grpc.ClientConn),
+		clients:     make(map[string]ptraceotlp.Client),
+		initialized: false,
+		mu:          &sync.RWMutex{},
+	}, nil
+}
+
+// Dial creates client connections and clients based on config.
+// Dial is expected to be called only once.
+func (f *Forwarder) Dial(ctx context.Context, opts ...grpc.DialOption) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.initialized {
+		return errors.New("already initialized")
+	}
+
+	connections := make(map[string]*grpc.ClientConn)
+	clients := make(map[string]ptraceotlp.Client, len(f.cfg.Endpoints))
+	for _, endpoint := range f.cfg.Endpoints {
+		client, conn, err := f.newTraceOTLPGRPCClientAndConn(ctx, endpoint, f.cfg.TLS, opts...)
+		if err != nil {
+			return fmt.Errorf("failed to create new trace otlp grpc client: %w", err)
+		}
+
+		connections[endpoint] = conn
+		clients[endpoint] = client
+	}
+
+	f.connections = connections
+	f.clients = clients
+	f.initialized = true
+
+	return nil
+}
+
+func (f *Forwarder) ForwardBatches(ctx context.Context, trace tempopb.Trace) error {
+	// tempopb.Trace is wire-compatible with OTLP, so we convert to bytes, and back.
+	m, err := trace.Marshal()
```

Comment:
we really need to avoid this marshal/unmarshal dance. we already do this once in the distributor and i've been trying to find a way to remove that for years.

since our trace is wire compatible with otlp we can just marshal it once when we send it and the other side should able to unmarshal.

Reply:
The problem is that the `Export()` method of `ptraceotlp.Client` accepts a `ptraceotlp.Request` as an argument, and I don't think there is a simple way to just convert `tempopb.Trace` type directly into the request type. I'd rather not do some reinterpret cast magic.

I don't know enough about Go, but in C++ such stuff would most likely be optimized away :D

Reply:
You can build the request with `ptraceotlp.NewRequestFromTraces(td)` if the conversion is moved up from the [shim](https://github.com/grafana/tempo/blob/main/modules/distributor/receiver/shim.go#L270-L283) to the distributor's `PushBatches`. The downside is that we would leak OTel to the distributor, but IMO it's worth for avoiding the conversion.

Reply:
I don't have a strong opinion on this. @joe-elliott which option would you prefer?

Reply:
> I'd rather not do some reinterpret cast magic.
100% agreed

I like @mapno's solution here.